### PR TITLE
lti/v1/login route and controller

### PR DESCRIPTION
**Pull request by nicklathe**
This PR adds the first LTI route and controller, which handle the initial LTI launch (Login) request coming from an LTI Platform (LMS). See [spec here](https://docs.google.com/document/d/1tTwAF40cnn2oUvK2CmRO-a5qYXkCR5rR5gBJS2-zV_U/edit#heading=h.emvwjfkaqwzq) for details on what this Login Controller should do.

## Testing story

```
rails test test/controllers/lti_v1_controller_test.rb
Running via Spring preloader in process 24143
Started with run options --seed 44741

  8/8: [=============================================================================================] 100% Time: 00:00:10, Time: 00:00:10

Finished in 14.24608s
8 tests, 18 assertions, 0 failures, 0 errors, 0 skips
```

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/test/controllers/lti_v1_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    assert_not_nil cookies[:state]
+    assert_not_nil cookies[:nonce]
+    assert_equal cookies[:state].length, 10
+    assert_equal cookies[:nonce].length, 10
+  end
+
+  test "given valid parameters, redirect URL should have valid auth request params" do
+    login_hint = "hint"
+    get "/lti/v1/login", params: {client_id: @integration.client_id, iss: @integration.issuer, login_hint: login_hint}
+    parsed_url = Rack::Utils.parse_query(@response.redirect_url.split("?")[1]).symbolize_keys
```
Comment (nicklathe) [inline dashboard/test/controllers/lti_v1_controller_test.rb]: This feels clunky, there has to be some kind of key on the `response` object that gives me just the query, but I haven't found it.

File: dashboard/app/controllers/lti_v1_controller.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      query_params = {platform_id: params[:platform_id]}
+    else
+      return unauthorized_status
+    end
+
+    begin
+      lti_integration = LtiIntegration.find_by!(query_params)
+    rescue ActiveRecord::RecordNotFound
+      return unauthorized_status
+    end
```
Comment (artem-vavilov) [inline dashboard/app/controllers/lti_v1_controller.rb]: Just curious, what is the reason for raising `ActiveRecord::RecordNotFound` error instead of checking for nil?
``` suggestion
    lti_integration = LtiIntegration.find_by(query_params)
    return unauthorized_status unless lti_integration
```

File: dashboard/app/controllers/lti_v1_controller.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      state: create_and_set_state,
+      response_mode: "form_post",
+      nonce: create_and_set_nonce,
+      prompt: "none",
+    }
+
+    auth_request_params = auth_request_body.to_query
+    auth_redirect_url = lti_integration[:auth_redirect_url]
+
+    redirect_to "#{auth_redirect_url}?#{auth_request_params}"
```
Comment (artem-vavilov) [inline dashboard/app/controllers/lti_v1_controller.rb]: Just a recomendation
```suggestion
    auth_redirect_url = URI(lti_integration[:auth_redirect_url])

    auth_redirect_url.query = {
      scope: "openid",
      response_type: "id_token",
      client_id: lti_integration[:client_id],
      redirect_uri: "https://studio.code.org/lti/v1/authenticate", # TODO: where can we configure this to be dynamic depending on deployment environment, i.e. prod, localhost, etc?
      login_hint:  params[:login_hint],
      lti_message_hint: params[:lti_message_hint] ? params[:lti_message_hint] : "", # TODO: is this needed? Canvas sends it, what happens if we don't send it back? What about other LTI Platforms?
      state: create_and_set_state,
      response_mode: "form_post",
      nonce: create_and_set_nonce,
      prompt: "none",
    }.to_query

    redirect_to auth_redirect_url
```

File: dashboard/app/controllers/lti_v1_controller.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    cookies[:"#{name}"] = {
+      value: value,
+      httponly: true,
+      same_site: :strict,
+    }
+  end
+
+  private def generate_random_string(length)
+    SecureRandom.alphanumeric length
+  end
```
Comment (artem-vavilov) [inline dashboard/app/controllers/lti_v1_controller.rb]: In the Ruby community, it is usually more acceptable not to declare each private method separately but to keep everything under the `private` block
https://github.com/rubocop/ruby-style-guide#access-modifiers-indentation
```suggestion
  private 

  def unauthorized_status
    render(status: :unauthorized, json: {error: "Unauthorized"})
  end

  def create_and_set_nonce
    # generate nonce, this will be set as a cookie.
    nonce = generate_random_string 10
    # generate hashed nonce, this will be returned to the LTI Platform in the
    # request params https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes
    hashed_nonce = Digest::SHA2.hexdigest nonce
    set_cookie("nonce", nonce)

    return hashed_nonce
  end

  def create_and_set_state
    state = generate_random_string 10
    set_cookie("state", state)

    return state
  end

  def set_cookie(name, value)
    cookies[:"#{name}"] = {
      value: value,
      httponly: true,
      same_site: :strict,
    }
  end

  def generate_random_string(length)
    SecureRandom.alphanumeric length
  end
```

File: dashboard/app/controllers/lti_v1_controller.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    end
+
+    begin
+      lti_integration = LtiIntegration.find_by!(query_params)
+    rescue ActiveRecord::RecordNotFound
+      return unauthorized_status
+    end
+
+    auth_request_body = {
+      scope: "openid",
```
Comment (artem-vavilov) [inline dashboard/app/controllers/lti_v1_controller.rb]: https://github.com/code-dot-org/code-dot-org/blob/staging/STYLEGUIDE.md#ruby-string-quotes

File: dashboard/app/controllers/lti_v1_controller.rb (hunk 2 above)
Comment (nicklathe) [reply dashboard/app/controllers/lti_v1_controller.rb]: I don't have a good reason, it seems like just checking for `nil` is the more straightforward way to handle this.

File: dashboard/app/controllers/lti_v1_controller.rb (hunk 5 above)
Comment (nicklathe) [reply dashboard/app/controllers/lti_v1_controller.rb]: Ah, yes I've seen single quotes and double quotes in the project, will change to single to stay standardized.

File: dashboard/app/controllers/lti_v1_controller.rb (hunk 3 above)
Comment (nicklathe) [reply dashboard/app/controllers/lti_v1_controller.rb]: Nice 👍

File: dashboard/app/controllers/lti_v1_controller.rb (hunk 4 above)
Comment (nicklathe) [reply dashboard/app/controllers/lti_v1_controller.rb]: That makes sense, I was wondering which way was the best, and saw another controller declaring each method as private, but I will go with what is preferable in the Ruby community.

File: dashboard/app/controllers/lti_v1_controller.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      response_type: "id_token",
+      client_id: lti_integration[:client_id],
+      redirect_uri: "https://studio.code.org/lti/v1/authenticate", # TODO: where can we configure this to be dynamic depending on deployment environment, i.e. prod, localhost, etc?
```
Comment (nicklathe) [inline dashboard/app/controllers/lti_v1_controller.rb]: Found the method I was looking for:
```
redirect_uri: CDO.studio_url('/lti/v1/authenticate', CDO.default_scheme),
```

File: dashboard/test/controllers/lti_v1_controller_test.rb (hunk 1 above)
Comment (nicklathe) [reply dashboard/test/controllers/lti_v1_controller_test.rb]: This might be cleaner `parsed_url = Rack::Utils.parse_query(URI(@response.redirect_url).query).symbolize_keys`

File: dashboard/app/controllers/lti_v1_controller.rb [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+      login_hint:  params[:login_hint],
+      lti_message_hint: params[:lti_message_hint] ? params[:lti_message_hint] : "", # TODO: is this needed? Canvas sends it, what happens if we don't send it back? What about other LTI Platforms?
```
Comment (artem-vavilov) [inline dashboard/app/controllers/lti_v1_controller.rb]: Since `nil.to_s == ''` we can simply do
```suggestion
      lti_message_hint: params[:lti_message_hint].to_s
```

File: dashboard/app/controllers/lti_v1_controller.rb (hunk 6 above)
Comment (artem-vavilov) [reply dashboard/app/controllers/lti_v1_controller.rb]: Does the `/lti/v1/authenticate` endpoint exist?

Comment (nicklathe) [reply dashboard/app/controllers/lti_v1_controller.rb]: Not yet. That's the next Jira item we'll work on.

**Comment by nicklathe** [comment]
Thanks @artem-vavilov for your feedback. I implemented the suggested changes.

File: dashboard/app/controllers/lti_v1_controller.rb [hunk 8]
```diff
@@ -0,0 +1,70 @@
+class LtiV1Controller < ApplicationController
+  # Don't require an authenticity token because LTI Platforms POST to this controller.
+  skip_before_action :verify_authenticity_token
+
+  # [GET/POST] /lti/v1/login(/:platform_id)
+  def login
+    if params[:client_id]
+      query_params = {client_id: params[:client_id], issuer: params[:iss]}
```
Comment (artem-vavilov) [inline dashboard/app/controllers/lti_v1_controller.rb:8]: Should the `iss` param be required when the `client_id` param is provided?

Comment (artem-vavilov) [inline dashboard/app/controllers/lti_v1_controller.rb:18]: Is it acceptable if only `client_id` parameter is provided without `iss` parameter?

Comment (nicklathe) [reply dashboard/app/controllers/lti_v1_controller.rb:18]: I'll add some light docs as comments in this controller, and a link to the LTI 1.3 spec. The `iss` parameter will always be present in the request, as it's required by LTI 1.3.

Comment (artem-vavilov) [reply dashboard/app/controllers/lti_v1_controller.rb:18]: I think we should add a check for this:
```suggestion
    if params[:client_id] && params[:iss]
      query_params = {client_id: params[:client_id], issuer: params[:iss]}
```
OR
```suggestion
    if params[:client_id]
      query_params = {client_id: params[:client_id], issuer: params.require(:iss)}
```

`params.require(:iss)` will raise `ActionController::ParameterMissing` if `iss` parameter isn't provided with `client_id` parameter

Comment (nicklathe) [reply dashboard/app/controllers/lti_v1_controller.rb:18]: We don't need to check for `iss`, it is required that an LTI 1.3 Platform include it in the Initiate Login request https://www.imsglobal.org/spec/security/v1p0/#step-1-third-party-initiated-login. If it's not present, that Platform isn't LTI 1.3 compliant, and we won't integrate with it.

`client_id` is not required, but most big LTI Platforms (LMS) will include this. With this logic, we check for `client_id` and if not present, fallback to `platform_id`, which is a unique ID we would have generated during the integration/onboarding step.

The reason we use `client_id` _and_ `iss` in our query is because a `client_id` is not guaranteed to be globally unique in our system (2 different LTI Platforms could generate the same id), however `platform_id` is guaranteed to be unique in our system, since we generate it.

File: dashboard/test/controllers/lti_v1_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    @integration = create :lti_integration
+  end
+
+  test 'given no params, return unauthorized' do
+    get '/lti/v1/login', params: {}
+    assert_response :unauthorized
+  end
+
+  test 'given a client_id that doesn not exist, return unauthorized' do
+    get '/lti/v1/login', params: {client_id: 'nope', iss: @integration.issuer}
```
Comment (elf-code) [inline dashboard/test/controllers/lti_v1_controller_test.rb]: This requires knowing that 'nope' is not the factory's client_id. Would an empty string also work?

File: dashboard/test/controllers/lti_v1_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    assert_response :unauthorized
+  end
+
+  test 'given a platform_id that does not exist, return unauthorized' do
+    get '/lti/v1/login/wrong-id', params: {}
+    assert_response :unauthorized
+  end
+
+  test 'given a valid client_id, return redirect' do
```
Comment (elf-code) [inline dashboard/test/controllers/lti_v1_controller_test.rb]: I would update the test string to include 'via GET' since the next test has the 'via POST'

File: dashboard/test/factories/factories.rb [hunk 11]
```diff
@@ -1691,7 +1691,7 @@
     issuer {"issuer"}
     client_id {"client_id"}
     platform_name {"platform_name"}
-    auth_redirect_url {"auth_redirect_url"}
+    auth_redirect_url {"http://test.org/auth"}
```
Comment (elf-code) [inline dashboard/test/factories/factories.rb:1694]: Was this change necessary or just stylistic update?

File: dashboard/app/controllers/lti_v1_controller.rb [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+  # [GET/POST] /lti/v1/login(/:platform_id)
+  #
+  # Most LTI Platforms should send a client_id as part of the login request.
+  # However, it is not required per the LTI 1.3 spec. In these cases, we can
+  # supply clients with a unique login URL, so we can identify the caller's
+  # identity. The requst will always contain the iss param, as required by the
+  # LTI 1.3 standard
+  # https://www.imsglobal.org/spec/security/v1p0/#step-1-third-party-initiated-login
+  def login
+    # First check if the request contains a client_id
```
Comment (elf-code) [inline dashboard/app/controllers/lti_v1_controller.rb]: I think these inline if / elsif / else comments are overkill and would rather remove them and let the code speak for itself. I like the link to the spec above and the longer comment.

File: dashboard/app/controllers/lti_v1_controller.rb [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    cookies[:"#{name}"] = {
+      value: value,
+      httponly: true,
+      same_site: :strict,
+    }
+  end
+
+  def generate_random_string(length)
+    SecureRandom.alphanumeric length
+  end
```
Comment (elf-code) [inline dashboard/app/controllers/lti_v1_controller.rb:79]: Do you think these will be re-used? It could be helpful to move them to a utils file. Maybe in the future

**Review by elf-code — APPROVED** [review]
Looks good! Comments are all stylistic nits, nothing is blocking.

File: dashboard/app/controllers/lti_v1_controller.rb (hunk 12 above)
Comment (nicklathe) [reply dashboard/app/controllers/lti_v1_controller.rb]: I agree. I think the long comment above explains what is needed.

File: dashboard/app/controllers/lti_v1_controller.rb (hunk 13 above)
Comment (nicklathe) [reply dashboard/app/controllers/lti_v1_controller.rb:79]: There isn't an existing helper function in the project (searched for `SecureRandom.alphanumeric`). My preference is just keep this here, and move it in the future _if_ some other functionality could benefit from using it.

File: dashboard/test/factories/factories.rb (hunk 11 above)
Comment (nicklathe) [reply dashboard/test/factories/factories.rb:1694]: Rails (or some custom logic) is prefixing the `auth_redirect_url` value unless there is a scheme in the string. 

Example with old string
```
# in factory.rb
auth_redirect_url {"auth_redirect_url"}

# log value of @response.redirect_url in unit test. See how auth_redirect_url string is getting prefixed with the local CDO domain.
https://test-studio.code.orgauth_redirect_url?client_id=client_id ...
```

Example with new string
```
# in factory.rb
auth_redirect_url {"http://test.org/auth"}

# log value of @response.redirect_url in unit test.
http://test.org/auth?client_id=client_id ...
```

This doesn't break any of the tests as they are right now, as I'm only testing the query string params of the redirect url.

File: dashboard/test/controllers/lti_v1_controller_test.rb (hunk 9 above)
Comment (nicklathe) [reply dashboard/test/controllers/lti_v1_controller_test.rb]: Good call, I'll change this to an empty string.
